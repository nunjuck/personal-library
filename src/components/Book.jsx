import React from "react";

const Book = () => {
  return (
    <article className="book-card">
      <div className="book-card__cover"></div>
      <h2 className="book-card__name">test</h2>
      <span className="book-card__availability"></span>
    </article>
  )
}

export default Book
