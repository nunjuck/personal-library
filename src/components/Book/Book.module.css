.card {
  position: relative;
  padding-bottom: 2rem;
  margin-bottom: 1rem;
  max-width: 320px;
}

.cover {
  img {
    width: 170px;
    height: 237px;
    outline: 1px solid #f2f2f2;
  }
}

.name {
  margin-top: 1rem;
  font-size: 1rem;
  color: var(--color-text-base);
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  line-height: 1.4;
  font-weight: normal;
}

.availability {
  display: block;
  margin-top: 0.6rem;
  color: var(--color-text-danger);
}

.availabilityYes {
  color: var(--color-text-access);
}

.copy {
  border: 0;
  background-color: transparent;
  padding: 0;
  color: var(--color-text-link);
  position: absolute;
  bottom: 0;
  cursor: pointer;

  @media (--screen-md) {
    font-size: 0.875rem;
    text-align: left;
  }
}
