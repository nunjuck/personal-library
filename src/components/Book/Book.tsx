import React, { FC } from 'react'

import { BookType } from '../../types/book'
import LazyImage from '../helpers/LazyImage'

import classes from './Book.module.css'

export const Book: FC<BookType> = ({ cover, title, availability }) => {
  const copySign = (data: string, event: React.FormEvent<HTMLButtonElement>) => {
    const th = event.currentTarget
    const name = th.innerText
    navigator.clipboard
      .writeText(data)
      .then(() => {
        th.innerText = 'Скопировано'
        setTimeout(() => {
          th.innerText = name
        }, 2000)
      })
      .catch((error) => {
        // eslint-disable-next-line no-console
        console.error('При копировании произошла ошибка. Обновите страницу или скопируйте вручную', error)
      })
  }

  return (
    <article className={classes.card}>
      <div className={classes.cover}>
        <LazyImage
          image={{
            alt: `Обложка книги ${title}`,
            src: cover,
            height: 237
          }}
        />
      </div>
      <h2 className={classes.name} title={title}>
        {title}
      </h2>
      <span className={`${classes.availability} ${availability ? classes.availabilityYes : ''}`}>
        {availability ? 'Доступна' : 'Занята'}
      </span>
      <button
        className={classes.copy}
        type='button'
        onClick={(event) => {
          copySign(title, event)
        }}
      >
        Скопировать название
      </button>
    </article>
  )
}
