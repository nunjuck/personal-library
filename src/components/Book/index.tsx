import React from "react";
import { BookType } from "../../types/book";
import LazyImage from "../helpers/LazyImage";

const Book = ({ cover, title, availability }: BookType) => {
  const copySign = (
    data: string,
    event: React.FormEvent<HTMLButtonElement>
  ) => {
    const th = event.currentTarget;
    const name = th.innerText;
    navigator.clipboard
      .writeText(data)
      .then(() => {
        th.innerText = "Скопировано";
        setTimeout(() => {
          th.innerText = name;
        }, 2000);
      })
      .catch((error) => {
        console.error(
          "При копировании произошла ошибка. Обновите страницу или скопируйте вручную",
          error
        );
      });
  };

  return (
    <article className="book-card">
      <div className="book-card__cover">
        <LazyImage
          image={{
            alt: `Обложка книги ${title}`,
            src: cover,
            height: 237,
          }}
        />
      </div>
      <h2 className="book-card__name" title={title}>
        {title}
      </h2>
      <span
        className={`book-card__availability ${
          availability ? "book-card__availability--yes" : ""
        }`}
      >
        {availability ? "Доступна" : "Занята"}
      </span>
      <button
        className="book-card__copy"
        type="button"
        onClick={(event) => {
          copySign(title, event);
        }}
      >
        Скопировать название
      </button>
    </article>
  );
};

export default Book;
