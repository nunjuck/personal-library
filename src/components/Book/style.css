.book-card {
  position: relative;
  padding-bottom: 2rem;
  margin-bottom: 1rem;
  max-width: 320px;
}

.book-card__cover {
  img {
    width: 170px;
    height: 237px;
    outline: 1px solid #F2F2F2;
  }
}

.book-card__name {
  margin-top: 1rem;
  font-size: 1rem;
  color: var(--color-text-base);
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  line-height: 1.4;
}

.book-card__availability {
  display: block;
  margin-top: 0.6rem;
  color: var(--color-text-danger);
  &--yes {
    color: var(--color-text-access);
  }
}

.book-card__copy {
  border: 0;
  background-color: transparent;
  padding: 0;
  color: var(--color-text-link);
  position: absolute;
  bottom: 0;
  cursor: pointer;

  @media (--screen-md) {
    font-size: 0.875rem;
    text-align: left;
  }
}
