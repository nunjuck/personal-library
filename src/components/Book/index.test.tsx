import React from "react";
import { render, screen } from "@testing-library/react";
import Book from ".";

const testBook = {
  cover: "image",
  title: "Чистый код",
  availability: true,
  id: "4546542124687845",
};

const setup = () => {
  const book = render(
    <Book
      cover={testBook.cover}
      title={testBook.title}
      availability={testBook.availability}
      id={testBook.id}
    />
  );

  const button = book.getByRole("button");
  const name = book.getByRole("heading");
  const cover = book.getByRole("img");

  return {
    button,
    name,
    cover,
    ...book,
  };
};

describe("Book", () => {
  test("render root element", () => {
    const book = setup();
    expect(book.getByRole("article")).toBeInTheDocument();
  });

  test("Book renders name", () => {
    const { name } = setup();
    expect(name).toBeInTheDocument();
  });

  test("Book renders cover", () => {
    const { cover } = setup();
    expect(cover).toBeInTheDocument();
  });

  test("Book renders with data", () => {
    const book = setup();
    expect(book).toMatchSnapshot();
  });

  test("Book renders button for copy name book", () => {
    const { button } = setup();
    expect(button).toBeInTheDocument();
    expect(button).toBeEnabled();
    expect(screen.getByText(/Скопировать/i)).toBeInTheDocument();
  });
});
