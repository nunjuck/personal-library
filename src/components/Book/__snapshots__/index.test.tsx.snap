// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Book renders with data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="book-card"
      >
        <div
          class="book-card__cover"
        >
          <span
            class=" lazy-load-image-background blur"
            style="background-image: url(image); background-size: 100% 100%; display: inline-block; height: 237px;"
          >
            <img
              alt="Обложка книги Чистый код"
              height="237"
              src="image"
            />
          </span>
        </div>
        <h2
          class="book-card__name"
        >
          Чистый код
        </h2>
        <span
          class="book-card__availability book-card__availability--yes"
        >
          Доступна
        </span>
        <button
          class="book-card__copy"
          type="button"
        >
          Скопировать название
        </button>
      </article>
    </div>
  </body>,
  "button": <button
    class="book-card__copy"
    type="button"
  >
    Скопировать название
  </button>,
  "container": <div>
    <article
      class="book-card"
    >
      <div
        class="book-card__cover"
      >
        <span
          class=" lazy-load-image-background blur"
          style="background-image: url(image); background-size: 100% 100%; display: inline-block; height: 237px;"
        >
          <img
            alt="Обложка книги Чистый код"
            height="237"
            src="image"
          />
        </span>
      </div>
      <h2
        class="book-card__name"
      >
        Чистый код
      </h2>
      <span
        class="book-card__availability book-card__availability--yes"
      >
        Доступна
      </span>
      <button
        class="book-card__copy"
        type="button"
      >
        Скопировать название
      </button>
    </article>
  </div>,
  "cover": <img
    alt="Обложка книги Чистый код"
    height="237"
    src="image"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "name": <h2
    class="book-card__name"
  >
    Чистый код
  </h2>,
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
